californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java

```diff
@@ -989,4 +1009,271 @@ protected CoapStack createUdpStack(NetworkConfig config, Outbox outbox) {
+		 *             defined
+		 * @see #bindAddress
+		 * @see #connector
+		 */
+		public CoapEndpointBuilder setConnector(Connector connector) {
+			if (this.bindAddress != null || this.connector != null) {
+				throw new IllegalArgumentException("bind address already defined!");
+			}
+			this.connector = connector;
+			this.applyConfiguration = false;
+
+			return this;
+		}
+
+		/**
+		 * Set connector to be configured and used by endpoint .
+		 * 
+		 * The {@link #bindAddress} could be defined at most once, so only one
+		 * setter of {@link #setPort(int)},
+		 * {@link #setInetSocketAddress(InetSocketAddress)}, or
+		 * {@link #setConnector(Connector)}, or
+		 * {@link #setConnectorWithAutoConfiguration(UDPConnector)} could be
+		 * used.
+		 * 
+		 * Provides a fluent API to chain setters.
+		 * 
+		 * @param connector connector to be used
+		 * @return this
+		 * @throws IllegalStateException, if {@link #bindAddress} is already
+		 *             defined
+		 * @throws IllegalArgumentException if applyConfiguration is
+		 *             {@code true}, but the connector is not a
+		 *             {@link UDPConnector}
+		 * @see #bindAddress
+		 * @see #connector
+		 */
+		public CoapEndpointBuilder setConnectorWithAutoConfiguration(UDPConnector connector) {
+			if (this.bindAddress != null || this.connector != null) {
+				throw new IllegalArgumentException("bind address already defined!");
+			}
+			this.connector = connector;
+
+			return this;
+		}
+
+		/**
+		 * Set observation store.
+		 * 
+		 * Provides a fluent API to chain setters.
+		 * 
+		 * @param store observation store
+		 * @return this
+		 * @see #observationStore
+		 */
+		public CoapEndpointBuilder setObservationStore(ObservationStore store) {
+			this.observationStore = store;
+			return this;
+		}
+
+		/**
+		 * Set message exchange store.
+		 * 
+		 * Provides a fluent API to chain setters.
+		 * 
+		 * @param exchangeStore message exchange store
+		 * @return this
+		 * @see #exchangeStore
+		 */
+		public CoapEndpointBuilder setMessageExchangeStore(MessageExchangeStore exchangeStore) {
+			this.exchangeStore = exchangeStore;
+			return this;
+		}
+
+		/**
+		 * Set endpoint context matcher.
+		 * 
+		 * Provides a fluent API to chain setters.
+		 * 
+		 * @param endpointContextMatcher endpoint context matcher
+		 * @return this
+		 * @see #endpointContextMatcher
+		 */
+		public CoapEndpointBuilder setEndpointContextMatcher(EndpointContextMatcher endpointContextMatcher) {
+			this.endpointContextMatcher = endpointContextMatcher;
+			return this;
+		}
+
+		/**
+		 * Create {@link CoapEndpoint} using the provided parameter or defaults.
+		 * 
+		 * @return new endpoint
+		 * @throws IllegalArgumentException if applyConfiguration is
+		 *             {@code true}, but the connector doesn't implement
+		 *             {@link ConnectorConfigurator}
+		 */
+		public CoapEndpoint build() {
+			if (config == null) {
+				config = NetworkConfig.getStandard();
```

Comment:
I really don't like this idea of standard config with a shared static instance. I already faced several issue about that.

I don't know if this has big impact but I would prefer to go with a new `NetworkConfig()` by default (which uses same default value without this shared instance and file creation)

Reply:
I'm not sure about the issues. I would guess, they are caused by "parallel initialization" or "adjust the configuration within the code". I already tried to isolate the usage of that `getStandard()` much more (see PR #502). 

But currently I don't think we should change the behaviour here too fundamentally (at least for now, so not within this PR).

Reply:
> I would guess, they are caused by "parallel initialization" or "adjust the configuration within the code".

You assume right. If I remember well the issue was 2 tests using NetworkConfig.getStandard() with different values launched in parallel.

In Leshan we don't use it any more.